### PR TITLE
gh-92869: ctypes: Add c_time_t

### PR DESCRIPTION
**Pull request by thp**
This implements a simple way of using `ctypes.c_time_t` for 32-bit and 64-bit `time_t` sizes. See #92869 for a discussion on why this is useful and considerations for different platforms. This doesn't take into account platforms where `time_t` is defined as `double` or anything else that's not a signed integer.

* Issue: gh-92869

### CODE REVIEW

**Comment by thp** [comment]
Bump after 24 days of no activity. Any takers for reviewing this PR?

**Comment by m-tmatma** [comment]
@thp 
It seems azure pipeline was failed.
You should re-trigger it.

File: Lib/ctypes/__init__.py
Comment (m-tmatma) [inline Lib/ctypes/__init__.py:573]: @thp 
Is it ok not to define else case?

**Review by m-tmatma — APPROVED** [review]

File: Lib/ctypes/__init__.py
Comment (thp) [reply Lib/ctypes/__init__.py:573]: If `time_t` is not 32-bit or 64-bit, `ctypes.c_time_t` will not be defined. I haven't seen a system where it's not 32-bit or 64-bit, but if we encounter such a system, the right type mapping can be added then.

Comment (m-tmatma) [reply Lib/ctypes/__init__.py:573]: @thp  
I think you should add to raise an exception on else case to make sure.
It may not happen, but it is a good practice to consider such exceptional case.

Comment (thp) [reply Lib/ctypes/__init__.py:573]: PR updated, now raises an exception.

File: Doc/library/ctypes.rst
Comment (AA-Turner) [inline Doc/library/ctypes.rst]:
```suggestion
   .. versionadded:: 3.12
```

File: Lib/ctypes/__init__.py
Comment (AA-Turner) [inline Lib/ctypes/__init__.py]:
```suggestion
    raise SystemError(f"Unexpected sizeof(time_t): {SIZEOF_TIME_T:d}")
```

File: Misc/NEWS.d/next/Library/2022-05-17-06-27-39.gh-issue-92869.t8oBkw.rst
Comment (AA-Turner) [inline Misc/NEWS.d/next/Library/2022-05-17-06-27-39.gh-issue-92869.t8oBkw.rst]:
```suggestion
Added :class:`~ctypes.c_time_t` to :mod:`ctypes`, which has the same size as
the :c:type:`time_t` type in C (either 32-bit or 64-bit)
```

**Review by AA-Turner — COMMENTED** [review]
Seems reasonable, please add tests though.

A

**Comment by AA-Turner** [comment]
Pinging listed names from the experts index for review.

A

File: Lib/ctypes/__init__.py
Comment (m-tmatma) [reply Lib/ctypes/__init__.py:573]: It seems I had approved this PR yesterday before the PR was updated,
But it was a mistake. I made it on a smart phone.

It seem I can't dismiss  the approval for this repository.

File: Lib/ctypes/test/test_sizes.py
Comment (thp) [inline Lib/ctypes/test/test_sizes.py]: Simple test case added.

As the reason for `c_time_t` is mostly to get the size of the object out of it, `test_sizes.py` is probably the best place to test for this. This test is mostly about checking for the existence of the field, and that it's 4 or 8. If/when platforms are found that have a differently-sized `c_time_t`, this test can be updated.

In the future, if we find a platform where `c_time_t` is of a different type (different size or even non-integer type), some other tests could be added.

**Comment by thp** [comment]
Found that the ctypes documentation's "Calling functions" documentation does this (without setting `restype`):

```
>>> print(libc.time(None))  
1150640792
```

This uses the default return type of `int`, which is wrong on systems that use 64-bit `time_t` (and just a happy coincidence on systems that use 32-bit `time_t` and happen to have 32-bit `int`):

```
>>> from ctypes import *
>>> libc = CDLL("libc.so.6")
>>> libc.time
<_FuncPtr object at 0x...>
>>> libc.time.restype
<class 'ctypes.c_int'>
```

Updated the documentation accordingly with setting `libc.time.restype` to the newly-added `c_time_t`, even though this introduces the concept of `restype` and `argtypes` quite early, but it's probably for the best, as not doing it can result in wrong return types as seen here. Alternatively, another libc function could be used for the "Calling functions" section, and `libc.time` used as an example for the "Return types" section.

The `libc.time.restype == c_int` error will be noticeable from 2038 onwards due to 32-bit overflow, at least on systems that have 64-bit `time_t` and 32-bit `int` (all modern systems?).

File: Lib/ctypes/test/test_sizes.py
Comment (m-tmatma) [reply Lib/ctypes/test/test_sizes.py]: Why don't you compare `sizeof(c_time_t)` with `SIZEOF_TIME_T`?

Comment (AA-Turner) [reply Lib/ctypes/test/test_sizes.py]: It seems like you should be able to use an independent version from `_testcapi`.

A

**Comment by AA-Turner** [comment]
> Alternatively, another libc function could be used for the "Calling functions" section, and libc.time used as an example for the "Return types" section.

I think this is a better approach. (Thanks for noticing!)

A

**Comment by thp** [comment]
> > Alternatively, another libc function could be used for the "Calling functions" section, and libc.time used as an example for the "Return types" section.
> 
> I think this is a better approach. (Thanks for noticing!)

Moved the `time()` function into the return type docs, and used `rand()` from libc instead (which conveniently has no arguments and returns a C `int`).

File: Lib/ctypes/test/test_sizes.py
Comment (thp) [reply Lib/ctypes/test/test_sizes.py]: (closing this conversation as resolved using `SIZEOF_TIME_T`, did not consider `_testcapi` at the moment)

**Comment by thp** [comment]
Merged upstream `main` branch, as there was a merge conflict shown in the Github Web UI.

Review/feedback/merge appreciated.

**Comment by thp** [comment]
Bump after 21 days of no feedback.

I know this isn't high priority, but maybe someone can review/approve this. Or just any outlook like "I won't have time to look into this before after summer".

If I should stop bumping this PR, please let me know - I hope the practice of waiting at least 20 days before nagging again is a good compromise between not being too annoying but still maybe getting it in front of the right people to review.

**Review by AA-Turner — APPROVED** [review]
This looks good from my perspective, but I'm not a ctypes specialist. I've asked if a core dev can review.

A

**Review by kumaraditya303 — APPROVED** [review]
LGTM

**Review by gpshead — APPROVED** [review]

**Comment by gpshead** [comment]
apologies about the commit message, i think i misattributed primary authorship in there by misreading github.  Thanks for the PR!

**Comment by AA-Turner** [comment]
To confirm this was all @thp's work! Thanks for the review & merge @gpshead.

A

**Pull request merged**
